docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,869 @@
+import MyWidget from './MyWidget';
+
+export default class NameHandler extends WidgetBase {
+	@watch() private currentName: string = 'Alice';
+
+	protected render() {
+		return (
+			<MyWidget
+				name={this.currentName}
+				onNameChange={(newName) => {
+					this.currentName = newName;
+				}}
+			/>
+		);
+	}
+}
+```
+
+## Advanced: abstracting and injecting state
+
+When implementing complex responsibilities, following a pattern of state encapsulation within widgets can result in bloated, unmanageable components. Another problem can arise in large applications where state is required in leaf widgets, but not in intermediate containers within the VDOM hierarchy. Passing state down such a complex widget hierarchy adds brittle, unnecessary code.
+
+Dojo provides the [Stores component](../stores/supplemental.md) to solve these issues by abstracting state management into its own dedicated context, then injecting relevant portions of the application's state into specific widgets that require it.
+
+# Widget development best practices
+
+When working with Dojo widgets, a few important principles should be kept in mind to avoid introducing anti-patterns into application code. Attempting to use the framework in an unsupported way can cause unexpected behavior and introduce difficult to find bugs into an application.
+
+## Widget properties
+
+-   Widgets should treat the properties passed to them as read-only.
+    -   Changes made to received properties within a widget cannot be propagated back to the framework, and result in a divergence between the widget and the framework.
+-   Widgets should avoid deriving further render state from their properties, and instead rely on their complete render state being provided to them.
+    -   Deriving render state can cause similar divergences between the widget and the framework as modifying received properties; the framework is not aware of the derived state, so cannot properly determine when a widget has been updated and requires invalidation and re-rendering.
+-   Internal or private state can be fully encapsulated within a widget, if required.
+    -   It is a valid and often desireable pattern to implement 'pure' widgets that incur no side effects and receive their entire state as properties, but this is not the only pattern in Dojo widget development.
+
+## Using class-based widgets
+
+-   The _`__render__`_, _`__setProperties__`_, and _`__setChildren__`_ functions are internal framework implementation details and should never be called nor overridden in application code.
+-   Applications should not instantiate widgets directly - Dojo fully manages the lifecycle of widget instances, including instantiation, caching and destruction.
+    -   Applications only need to provide widgets to the framework as generic type parameters when calling [the `w()` method](#instantiating-vdom-nodes), or implicitly via [TSX tags](#tsx-widget-example), as VDOM output from widgets' render functions.
```

Comment:
"Applications only need to provide widgets to the framework as generic type parameters when calling [the `w()` method](#instantiating-vdom-nodes), or implicitly via [TSX tags](#tsx-widget-example), as VDOM output from widgets' render functions."

I'm not sure I understand what this means?

Reply:
Convoluted way of saying refer to widgets as generic type arguments, not as concrete instances. I've removed this sub point as the main one is a lot clearer.